verilog/CST/net.cc

```diff
@@ -29,4 +33,13 @@ std::vector<verible::TreeSearchMatch> FindAllNetDeclarations(
   return SearchSyntaxTree(root, NodekNetDeclaration());
 }
 
+const SyntaxTreeLeaf* GetIdentifierFromNetDeclaration(
+    const Symbol& symbol) {
+  const auto* data_type =
+      verible::GetSubtreeAsSymbol(symbol, NodeEnum::kNetDeclaration, 1);
+  const auto* identifier_symbol =
+      verible::GetSubtreeAsSymbol(*data_type, NodeEnum::kDataTypeImplicitIdDimensions, 1);
+  return AutoUnwrapIdentifier(*ABSL_DIE_IF_NULL(identifier_symbol));
+}
```

Comment:
This is implemented to mimic what is done for port declarations. Is that acceptable?

Also, what to do for data declaration? Would that be ok to add `verilog/CST/data.cc` for that purpose? Or should that be done by hand in the actual lint rule class?

Reply:
Yes, we basically need more and more extractor functions like this to isolate the implementation details of the CST, which can be prone to internal change and restructuring.
CST/data.cc for data declarations would be fine.  If you see opportunities in verilog.y to improve structures or make them more consistent, that is welcome too, we just have to keep the tests passing.

Reply:
@fangism I actually have a question about the tree for the Net nodes.

Consider the following example:
```
wire A, B, C;
```

It produces:
```
      Node (tag: kNetDeclaration) {     
        (#"wire" @310-314: "wire")        
        Node (tag: kDataTypeImplicitIdDimensions) {
          Node (tag: kDataType) {         
            Node (tag: kPackedDimensions) {  
            }  
          }  
          (#SymbolIdentifier @315-316: "A")
          Node (tag: kUnpackedDimensions) {              
          }                                
        }                                      
        (#',' @316-317: ",")                   
        Node (tag: kNetVariableDeclarationAssign) {
          Node (tag: kNetVariable) {
            (#SymbolIdentifier @318-319: "B")
            Node (tag: kUnpackedDimensions) {
            }               
          }
          (#',' @319-320: ",")      
          Node (tag: kNetVariable) {    
            (#SymbolIdentifier @321-322: "C")
            Node (tag: kUnpackedDimensions) {
            }                             
          }                                  
        }      
        (#';' @322-323: ";")
      }
```

It doesn't look correct.

Shouldn't this create a separate `kNetDeclaration` nodes for each net?
Or maybe an additional tag for a *list* of nets, like it is done for `kDataDeclaration`, where:
```
logic X, Y, Z;
```
produces:
```
      Node (tag: kDataDeclaration) {
        Node (tag: kInstantiationBase) {
          Node (tag: kInstantiationType) {
            Node (tag: kDataTypePrimitive) {
              (#"logic" @350-355: "logic")
              Node (tag: kPackedDimensions) {
              }
            }
          }
          Node (tag: kGateInstanceRegisterVariableList) {
            Node (tag: kRegisterVariable) {
              (#SymbolIdentifier @356-357: "X")
              Node (tag: kUnpackedDimensions) {
              }
            }
            (#',' @357-358: ",")
            Node (tag: kRegisterVariable) {
              (#SymbolIdentifier @359-360: "Y")
              Node (tag: kUnpackedDimensions) {
              }
            }
            (#',' @360-361: ",")
            Node (tag: kRegisterVariable) {
              (#SymbolIdentifier @362-363: "Z")
              Node (tag: kUnpackedDimensions) {
              }
            }
          }
        }
        (#';' @363-364: ";")
```
?

BTW. I can see that you can still pretty much extract all the information you need from the tree, e.g:
```
wire A = 1, B = 0, C = 1;
```

Gives:
```
      Node (tag: kNetDeclaration) {        
        (#"wire" @310-314: "wire")             
        Node (tag: kDataTypeImplicitIdDimensions) {
          Node (tag: kDataType) {
            Node (tag: kPackedDimensions) {
            }                   
          }                                
          (#SymbolIdentifier @315-316: "A")    
          Node (tag: kUnpackedDimensions) {    
          }    
        }    
        Node (tag: kTrailingAssign) {
          (#'=' @317-318: "=")
          Node (tag: kExpression) {
            Node (tag: kNumber) {
              (#TK_DecNumber @319-320: "1")
            }                           
          }                               
        }                                   
        (#',' @320-321: ",")              
        Node (tag: kNetVariableDeclarationAssign) {
          Node (tag: kNetDeclarationAssignment) {
            (#SymbolIdentifier @322-323: "B")
            (#'=' @324-325: "=")
            Node (tag: kExpression) {                    
              Node (tag: kNumber) {        
                (#TK_DecNumber @326-327: "0")  
              }                                
            }  
          }  
          (#',' @327-328: ",")
          Node (tag: kNetDeclarationAssignment) {
            (#SymbolIdentifier @329-330: "C")
            (#'=' @331-332: "=")
            Node (tag: kExpression) {
              Node (tag: kNumber) {     
                (#TK_DecNumber @333-334: "1")
              }                             
            }                             
          }                                  
        }      
        (#';' @334-335: ";")
```

But I just don't see why `A` should be represented differently than `B`, and `C` in the example above.

Is that a bug, or am I missing something?

Reply:
All good observations.  The CST structure was very ad hoc, and contains inconsistencies like this -- they were done quickly to make sure the information was captured *somewhere* in some form, but the structure isn't as consistent as it could be.  You may take the liberty to re-pack these structures (e.g. around kNetDeclaration) into a better list form in verilog.y.  You shouldn't have to change the grammar rules, just their semantic actions, to get a structure that makes better sense (e.g. make A look like B and C).  If you're confident in playing with the grammar, there may be an opportunity to rewrite it so that A simply becomes a kNetVariableDeclarationAssign.